stanford_soe_r25.module

```diff
@@ -0,0 +1,889 @@
+              }
+            }
+
+            // get the reservation form and render it into the block
+            $form = drupal_get_form("stanford_soe_r25_reservation", $room_id, $capacity, $external_display, $can_book['external_acct']);
+            $block['subject'] = t("Request") . " " . $rooms[$room_id]['display_name'];
+            $block['content'] = drupal_render($form);
+            // otherwise, if this room authenticates externally, get the auth link from the external auth module
+          }
+          else {
+            // the room is not reservable. whatever shall we do?
+
+            // if the room can be authenticated externally, display a login link if we can
+            if (!empty($rooms[$room_id]['authentication']) &&
+              $rooms[$room_id]['authentication'] > stanford_soe_r25_AUTH_DRUPAL_ACCT
+            ) {
+              // present the user with an authentication link to access the room reservation form
+              // but if no module implementing the hook was found we need to put up a message
+              if (empty($can_book['external_module'])) {
+                $block['subject'] = '';
+                $block['content'] = 'This room requires external authentication, but no login link was found.';
+              }
+              else {
+                $block['subject'] = '';
+                $login_msg = variable_get('stanford_soe_r25_ext_login_msg', 'Authenticate');
+                $block['content'] = module_invoke($can_book['external_module'], 'stanford_soe_r25_external_link', $login_msg);
+              }
+            }
+            else {
+              // otherwise, access to reserve room is based on Drupal permissions so offer a login link if
+              // user is anonymous, or a 'no access' message if user is logged in
+
+              // if the user has no reservation permission and is logged in, then print a "no access" message
+              // otherwise the user is anonymous so display the login link,
+              if (user_is_logged_in()) {
+                $no_access = variable_get('stanford_soe_r25_notpermitted_msg', array(
+                  'value' => '',
+                  'format' => NULL
+                ));
+                if (!empty($no_access)) {
+                  $block['subject'] = '';
+                  $block['content'] = check_markup($no_access['value'], $no_access['format']);
+                }
+              }
+              else {
+                $login_msg = variable_get('stanford_soe_r25_login_msg', '');
+                $login_uri = variable_get('stanford_soe_r25_login_uri', '');
+                if (!empty($login_msg) && !empty($login_uri)) {
+                  $block['subject'] = '';
+                  // the login link uses a button CSS class that you may define in a css file.
+                  $block['content'] = l($login_msg, $login_uri, array(
+                    'query' => array('destination' => current_path() . '#stanford-r25-reservation'),
```

Comment:
current_path() values are not sanitized and should not be used directly. This could introduce XSS attacks.

Reply:
actually, this is used in the l() function and is probably fine. I am not sure if post drupal 7.58, that the destination parameter is sanitized or not. Worth investigation.